### ISSUE TITLE
Roadie *SHOULD* extract external stylesheets by default (especially when using a CDN)

### ISSUE THREAD

**Issue by dimroc** [Mange/roadie#107]
I have a rails app that uses Fastly CDN to distribute all content, including those in an email. To support Roadie, I made a simple custom asset provider you can see here: https://gist.github.com/dimroc/eeb519c36152dbc133ae

By using a CDN in both the website and the emails, all of my .css files are external. Why you may ask? Because I wanted parity between my email previews and my actual emails.

With the change in e608038117d860e4268d45d42ff14be944edc822, roadie stops extracting external css files by default, therefore breaking my emails :-1: . I don't think this should be the case, and we should instead rely on data-roadie-ignore. But that flies in the face of #106.

I propose reverting e608038117d860e4268d45d42ff14be944edc822, which I've done in my fork, but perhaps there's a better way out of this. Let me know your thoughts @Mange.

Thanks for a great gem!

**Comment by Mange** [comment]
Thank you for reporting this. The behavior you used was always a bug from
the project's perspective, so I will not revert the change.
I will, however, implement your use case, enabling external assets to be
found through asset providers.
I don't know exactly how I'll do it yet (configuration, another method on
the asset providers, etc) but I will add it so you should be able to
upgrade to that future version and get everything working again.

I will not implement external assets out of the box as the way to download
and cache those assets could be highly dependent on the application, and I
don't want network handling added to Roadie. It's the perfect example of
when it's important to support a flexible API to find assets, though.

I'm sorry I broke your app. I hope my rather long (and redundant) answer
convince you that I won't leave it at this state.

Any extra input you might have on how you'd want this feature to work is
most welcome. I'll read your provider's code tomorrow and gain that input,
at least. I will also check out your fork.

tors 7 maj 2015 16:28 Dimitri Roche <notifications@github.com> skrev:

**Comment by dimroc** [comment]
Thanks a lot @Mange. My only other input would be to look at the environments/<env>.rb files and utilize config.action_mailer.asset_host.

For example, I configure my CDN assets in environments/production.rb like below:

```ruby
  config.action_mailer.asset_host = ENV['FASTLY_CDN_URL']
```

**Comment by Mange** [comment]
Asset host is a bit more complex as it can be a string, an array of strings
or a callable.

sön 10 maj 2015 15:24 Dimitri Roche <notifications@github.com> skrev:

**Comment by Mange** [comment]
I just released version 3.1.0.rc1 with support for downloading external stylesheets. Let me know how it works for you.

I'm going to release a RC for a new version of `roadie-rails` too if you are using this with Rails.

Please reopen this issue if you cannot get it to work or need assistance. The README has a few examples.

**Comment by dimroc** [comment]
Ok, will do. Thanks.

**Comment by dimroc** [comment]
Hi @Mange,
Ran into the following issue:

Adding `loopandtie-qa-herokuapp-com.global.ssl.fastly.net` to the whitelist like so:

```ruby
provider = Roadie::NetHttpProvider.new(whitelist: Array.wrap("loopandtie-qa-herokuapp-com.global.ssl.fastly.net"))
options = { external_asset_providers: [provider] }
....
```

Yields the following error:

```
GiftMailer#gift_email: processed outbound mail in 512.9ms
NoMethodError: undefined method `request_uri' for #<URI::Generic:0x007fec12222da0>
	from /app/vendor/ruby-2.2.3/lib/ruby/2.2.0/net/http/generic_request.rb:19:in `initialize'
	from /app/vendor/ruby-2.2.3/lib/ruby/2.2.0/net/http/request.rb:14:in `initialize'
	from /app/vendor/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:1285:in `new'
	from /app/vendor/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:1285:in `request_get'
	from /app/vendor/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:480:in `block in get_response'
	from /app/vendor/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:853:in `start'
	from /app/vendor/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:583:in `start'
	from /app/vendor/ruby-2.2.3/lib/ruby/2.2.0/net/http.rb:478:in `get_response'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/net_http_provider.rb:52:in `download'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/net_http_provider.rb:35:in `find_stylesheet!'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/provider_list.rb:57:in `block in find_stylesheet!'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/provider_list.rb:55:in `each'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/provider_list.rb:55:in `find_stylesheet!'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/asset_scanner.rb:88:in `read_link_element'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/asset_scanner.rb:78:in `read_stylesheet'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/asset_scanner.rb:45:in `block in extract_css'
```

A quick google leads me to believer it's because my url doesn't start with http:// or https:// but switching to that bombs because the asset pipeline doesn't cooperate and lead with the protocol:

```
GiftMailer#gift_email: processed outbound mail in 449.2ms
Roadie::ProvidersFailed: Could not find stylesheet "https://loopandtie-qa-herokuapp-com.global.ssl.fastly.net/assets/email/email-eb3a258872a771ef735461ed7d7f527a.css": All providers failed
Used providers:
	#<Roadie::NetHttpProvider whitelist: #<Set: {"https://loopandtie-qa-herokuapp-com.global.ssl.fastly.net"}>>: loopandtie-qa-herokuapp-com.global.ssl.fastly.net is not part of whitelist!

	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/provider_list.rb:62:in `find_stylesheet!'
	from /app/vendor/bundle/ruby/2.2.0/gems/roadie-3.1.0.rc1/lib/roadie/asset_scanner.rb:88:in `read_link_element'
```

Note the discrepancy in the URLs. Further reading: http://stackoverflow.com/questions/2291919/while-loading-webpage-undefined-method-request-uri-for

I'll continue to mess around and force my URLs to have a leading http/https, but I think we should handle this without blowing up.

**Comment by dimroc** [comment]
PS. I can't reopen the issue as I don't have permissions.

**Comment by dimroc** [comment]
ruby 2.2.3, rails 4.2.4

**Comment by dimroc** [comment]
I have had paths being sent to roadie looking like


1. `//loopandtie-qa-herokuapp-com.global.ssl.fastly.net/assets/email/email-eb3a258872a771ef735461ed7d7f527a.css`
2. `loopandtie-qa-herokuapp-com.global.ssl.fastly.net/assets/email/email-eb3a258872a771ef735461ed7d7f527a.css`

And I believe it's the result of the asset pipeline but I'll look into it. Removing all whitelist, which in turn allows all domains, unblocked me.

**Comment by dimroc** [comment]
My custom provider just assumed `https`:

```ruby
class FastlyAssetsProvider
  include Roadie::AssetProvider

  def find_stylesheet(name)
    Rails.logger.info("## FastlyAssetsProvider retrieving asset named: #{name}")
    asset_host = Rails.application.config.action_mailer.asset_host
    return nil unless asset_host.present?

    uri = URI(name)
    uri.scheme = 'https'
    response = URI.parse(uri.to_s).read

    Roadie::Stylesheet.new(name, response)
  rescue OpenURI::HTTPError
  rescue NoMethodError
    nil
  end
end
```

**Comment by Mange** [comment]
The whitelist is a list of *host* names. If you specify a valid URL (with a proper scheme/protocol, for example `https`), and add the hostname to the whitelist it should work.

I'll make sure `NetHttpProvider` raises an `ArgumentError` on broken whitelists to help people understand this. Thanks for the feedback.

I might also, just to make it easier for everyone, support the non-standard scheme-less `//` URLs. :-)

**Comment by Mange** [comment]
Latest `master` now complains about bad hostnames and supports scheme-less URLs. A new RC will be forthcoming soonish.

**Comment by Mange** [comment]
I just released 3.1.0. You are welcome to upgrade to it. :-)
